src/base/misc/fatfs.c

```diff
@@ -408,9 +408,43 @@ int read_data(fatfs_t *f, unsigned pos, unsigned char *buf)
   return read_cluster(f, pos / f->cluster_secs + 2, pos % f->cluster_secs, buf);
 }
 
+static int get_bpb_version(struct on_disk_bpb *bpb)
+{
+
+  if (bpb->v340_400_signature == BPB_SIG_V400)
+    return 400;
+
+  if (bpb->v340_400_signature == BPB_SIG_V340)
+    return 340;
+
+  if (bpb->num_sectors_small == 0) { // FAT16B
+    if (bpb->v300_320_hidden_sectors == bpb->v331_400_hidden_sectors) {
+      // We know the data following v300_320_hidden_sectors must be greater
+      // than zero if it's to represent num_sectors_large and so if uint16
+      // and uint32 representations of hidden_sectors are the same value we
+      // must have a uint32 field, so the BPB is v3.31
+      return 331;
+    } else {
+      // Since v3.00 doesn't support FAT16B it must be v3.20
+      return 320;
+    }
+  }
+
+  // FAT12 or FAT16
+  // We know the data following v300_320_hidden_sectors must be zero to
+  // represent num_sectors_large on a v320 or v331 BPB, but equally it could
+  // just be data in a v300 BPB, so we can't distinguish between BPB
+  // versions 3.00, 3.20, and 3.31.
+  return 300;
+}
+
 static void set_geometry(fatfs_t *f, unsigned char *b)
 {
   struct on_disk_bpb *bpb = (struct on_disk_bpb *) &b[0x0b];
+  int version;
```

Comment:
Could you please leave set_geometry()
as simple as possible and introduce another
function for parsing and converting boot.blk?
I simply don't understand why this stuff is needed
on a generic path: no one uses boot.blk, so
there is no need to complicate the generic path.